File: android/src/main/kotlin/lazebny/io/jc/platformView/JcConferencePlatformView.kt
package lazebny.io.jc.platformView

import android.content.Context
import android.view.ViewGroup
import android.widget.FrameLayout
import com.juphoon.cloud.JCMediaChannel
import com.juphoon.cloud.JCMediaDevice
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.StandardMessageCodec
import io.flutter.plugin.platform.PlatformView
import io.flutter.plugin.platform.PlatformViewFactory
import lazebny.io.jc.common.JcViewController
import lazebny.io.jc.logic.JcWrapper.JCManager
import lazebny.io.jc.utils.ViewHelper

class JcConferencePlatformViewFactory(private val messenger: BinaryMessenger) :
    PlatformViewFactory(StandardMessageCodec.INSTANCE) {

    override fun create(context: Context?, viewId: Int, args: Any?): PlatformView {
        val map = args as Map<*, *>
        return JcConferencePlatformView(context!!, messenger, viewId, map["uid"] as String)
    }
}

class JcConferencePlatformView(
    context: Context,
    private val messenger: BinaryMessenger,
    private val viewId: Int,
    uid: String,
) : PlatformView, JcViewController {

    private var view: FrameLayout

    init {

        val participant =
            JCManager.getInstance().mediaChannel.participants.find { it.userId == uid }
        val canvas = participant?.startVideo(
            JCMediaDevice.RENDER_FULL_SCREEN,
            JCMediaChannel.PICTURESIZE_MIN
        )
        view = FrameLayout(canvas!!.videoView.context)
        canvas.videoView.let {
            val parent = it.parent as ViewGroup?
            parent?.removeView(it)
        }
        view.addView(
            canvas.videoView,
            FrameLayout.LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT,
                ViewGroup.LayoutParams.WRAP_CONTENT
            ),
        )
        JcViewController.setUp(messenger, viewId, this)
    }

    override fun getView() = view

    override fun dispose() {
        view.removeAllViews()
        JcViewController.setUp(messenger, viewId, null)
    }

    override fun setLayoutParams(width: Double, height: Double) {
        view.layoutParams?.width = ViewHelper.convertDpToPixel(width.toFloat()).toInt()
        view.layoutParams?.height = ViewHelper.convertDpToPixel(height.toFloat()).toInt()
    }

}

File: ios/Classes/platformView/JcConferencePlatformView.swift
import Flutter
import JCSDKOC

class JcConferencePlatformViewFactory: NSObject, FlutterPlatformViewFactory {
    let messenger: FlutterBinaryMessenger
    
    init(messenger: FlutterBinaryMessenger) {
        self.messenger = messenger
        super.init()
    }
    
    func create(withFrame frame: CGRect, viewIdentifier viewId: Int64, arguments args: Any?) -> FlutterPlatformView {
        let argsMap = args as! [String: Any]
        let uid = argsMap["uid"] as! String
        return JcConferencePlatformView(messenger: messenger, viewId: viewId, uid: uid)
    }
    
    func createArgsCodec() -> FlutterMessageCodec & NSObjectProtocol {
        return FlutterStandardMessageCodec.sharedInstance()
    }
}

class JcConferencePlatformView: NSObject, FlutterPlatformView, JcViewController {
    let _view: UIView
    let viewId: Int64
    let messenger: FlutterBinaryMessenger
    let uid: String
    private var canvas: JCMediaDeviceVideoCanvas?
    
    init(messenger: FlutterBinaryMessenger, viewId: Int64, uid: String) {
        self.messenger = messenger
        self.viewId = viewId
        self.uid = uid
        self._view = UIView()
        super.init()
        
        let participant = JCRoom.shared.mediaChannel.participants.compactMap({ ($0 as! JCMediaChannelParticipant) }).first { $0.userId == uid }
        canvas = participant?.startVideo(.fullScreen, pictureSize: .large)
        
        if let videoView = canvas?.videoView {
            _view.addSubview(videoView)
        }
        
        JcViewControllerSetup.setUp(binaryMessenger: messenger, viewId: viewId, viewController: self)
    }
    
    func view() -> UIView {
        return _view
    }
    
    func dispose() {
        canvas?.videoView.removeFromSuperview()
        JcViewControllerSetup.setUp(binaryMessenger: messenger, viewId: viewId, viewController: nil)
    }
    
    func setLayoutParams(width: Double, height: Double) {
        canvas?.videoView.frame = CGRect(x: 0, y: 0, width: width, height: height)
    }
}
